Parent resource of [.../target/classes/com/example/application/views/MyView.class] not found in the resources!

### Description of the bug

If you take an empty project, e.g. from start.spring.io, start it and add a Java class in a sub package, e.g. `views/MyView.java` then Spring Boot restarts and fails because of
```
Caused by: java.lang.IllegalStateException: Parent resource of [/.../target/classes/com/example/application/views/MyView.class] not found in the resources!
        at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.collectResources(VaadinServletContextInitializer.java:1022) ~[vaadin-spring-24.4-SNAPSHOT.jar:na]
        at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.getResources(VaadinServletContextInitializer.java:963) ~[vaadin-spring-24.4-SNAPSHOT.jar:na]
        at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:457) ~[spring-context-6.1.3.jar:6.1.3]
        at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.findCandidateComponents(ClassPathScanningCandidateComponentProvider.java:351) ~[spring-context-6.1.3.jar:6.1.3]
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[na:na]
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[na:na]
        at com.vaadin.flow.spring.VaadinServletContextInitializer$VaadinAppShellContextListener.failFastContextInitialized(VaadinServletContextInitializer.java:644) ~[vaadin-spring-24.4-SNAPSHOT.jar:na]
        at com.vaadin.flow.spring.VaadinServletContextInitializer$FailFastServletContextListener.contextInitialized(VaadinServletContextInitializer.java:195) ~[vaadin-spring-24.4-SNAPSHOT.jar:na]
```

### Expected behavior

It works

### Minimal reproducible example

https://github.com/Artur-/empty-vaadin-app

```
mvn
```

```
mkdir src/main/java/com/example/application/views
echo "package com.example.application.views; public class MyView{}" >  src/main/java/com/example/application/views/MyView.java
mvn compile
```

### Versions

- Vaadin / Flow version:
- Java version:
- OS version:
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

--- a/flow-tests/test-redeployment/pom.xml
+++ b/flow-tests/test-redeployment/pom.xml
@@ -103,6 +103,29 @@
                 <groupId>com.vaadin</groupId>
                 <artifactId>flow-maven-plugin</artifactId>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-clean-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>clean-test-files</id>
+                        <phase>initialize</phase>
+                        <goals>
+                            <goal>clean</goal>
+                        </goals>
+                        <configuration>
+                            <filesets>
+                                <fileset>
+                                    <directory>${project.basedir}/src/main/java/com/vaadin/flow/uitest/ui</directory>
+                                    <includes>
+                                        <include>reloadaddedviews/</include>
+                                    </includes>
+                                </fileset>
+                            </filesets>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
         </plugins>
     </build>
 

--- a/flow-tests/test-redeployment/src/test/java/com/vaadin/flow/uitest/ui/DevModeClassCacheIT.java
+++ b/flow-tests/test-redeployment/src/test/java/com/vaadin/flow/uitest/ui/DevModeClassCacheIT.java
@@ -15,21 +15,47 @@
  */
 package com.vaadin.flow.uitest.ui;
 
+import javax.tools.JavaCompiler;
+import javax.tools.ToolProvider;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.List;
 
 import net.jcip.annotations.NotThreadSafe;
+import org.apache.commons.io.FileUtils;
+import org.junit.After;
 import org.junit.Assert;
+import org.junit.Before;
 import org.junit.Test;
 import org.openqa.selenium.By;
+import org.openqa.selenium.StaleElementReferenceException;
 
 import com.vaadin.testbench.TestBenchElement;
 
 @NotThreadSafe
 public class DevModeClassCacheIT extends AbstractReloadIT {
 
+    private static final Path VIEW_PATH = Path.of("com", "vaadin", "flow",
+            "uitest", "ui", "reloadaddedviews");
+
     @Override
     protected String getTestPath() {
         return super.getTestPath().replace("/view", "");
+    }
+
+    @Before
+    @After
+    public void removeTemporaryView() throws IOException {
+        Path baseDir = new File(System.getProperty("user.dir", ".")).toPath();
+        FileUtils.deleteDirectory(baseDir.resolve(Path.of("target", "classes"))
+                .resolve(VIEW_PATH).toFile());
+        FileUtils
+                .deleteDirectory(baseDir.resolve(Path.of("src", "main", "java"))
+                        .resolve(VIEW_PATH).toFile());
     }
 
     @Test
@@ -53,5 +79,58 @@ public class DevModeClassCacheIT extends AbstractReloadIT {
         Assert.assertEquals("Unexpected cached route packages.",
                 "com.vaadin.flow.uitest.ui",
                 allSpans.get(5).getText().split(":")[1]);
+
+        // Ensure newly created classes are correctly added to the cache
+        createNewViewReloadAndWait();
+
+        waitForElementPresent(By.id("last-span"));
+
+        allSpans = $("span").all();
+
+        for (int i = 1; i < 5; i++) {
+            String[] value = allSpans.get(i).getText().split(":");
+            Assert.assertTrue("Expected " + value[0] + " to be greater than 0.",
+                    Integer.parseInt(value[1]) > 0);
+        }
+
+        Assert.assertEquals("Unexpected cached route packages.",
+                "com.vaadin.flow.uitest.ui,com.vaadin.flow.uitest.ui.reloadaddedviews",
+                allSpans.get(5).getText().split(":")[1]);
     }
+
+    // create class on the fly
+    protected void createNewViewReloadAndWait() {
+        String viewId = getViewId();
+
+        Path baseDir = new File(System.getProperty("user.dir", ".")).toPath();
+        Path outputPath = baseDir.resolve(Path.of("target", "classes"));
+        Path sourcePath = baseDir.resolve(Path.of("src", "main", "java"))
+                .resolve(VIEW_PATH);
+        Path sourceFile = sourcePath.resolve("MyView.java");
+        try {
+            Files.createDirectories(sourcePath);
+            Files.writeString(sourceFile, """
+                    package com.vaadin.flow.uitest.ui.reloadaddedviews;
+                    import com.vaadin.flow.component.html.Div;
+                    import com.vaadin.flow.router.Route;
+                    @Route("compiled-at-runtime")
+                    public class MyView extends Div {}
+                    """);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+        int result = compiler.run(null, null, null, "-d", outputPath.toString(),
+                "-sourcepath", sourcePath.toString(), sourceFile.toString());
+        Assert.assertEquals("Failed to compile " + sourceFile, 0, result);
+
+        waitUntil(driver -> {
+            try {
+                return !getViewId().equals(viewId);
+            } catch (StaleElementReferenceException e) {
+                return false;
+            }
+        });
+    }
+
 }

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -993,6 +993,11 @@ public class VaadinServletContextInitializer
 
                 if (devModeCachingEnabled && valid.contains(originalPath)) {
                     resourcesList.add(resource);
+                    // Restore root paths to ensure new resources are correctly
+                    // validate and cached after a reload
+                    if (originalPath.endsWith("/")) {
+                        rootPaths.add(originalPath);
+                    }
                 } else {
                     if (path.endsWith(".jar!/")) {
                         resourcesList.add(resource);

